File: contracts/astrozap/src/math.rs
use num_bigint::BigInt;

/// The maximum number of iterations to do when solving the quadratic equation
const MAX_ITERATIONS: usize = 32;

/// Commission rate of the Astroport XYK pool, set to 0.3%
///
/// We can technically query the factory contract for this number, but this is, in my opinion,
/// unnecessary and a waste of gas because the rate is almost never going to change. If it does
/// change, we can always update this constant here and migrate the contract.
const COMMISSION_RATE_BPS: u64 = 30;

/// Equation describing the relation between the optimal swap amount (x) and the asset amounts. It
/// is a quadratic equation of the form `a * x^2 + b * x + c = 0` where `a, b, c >= 0`. For details,
/// see the document `docs/astrozap.pdf`
///
/// NOTE: Do we really need Uint512 or will Uint256 do? The maximum value we may reach is
///   (2^128 - 1)^3 ~= 1e+115
/// while the maximum amount Decimal256 supports is
///   2^256 - 1 ~= 1e+77
/// This being said, overflow should be very rare unless for pools with ultra-low-price shitcoins,
/// and even in such cases we will safely return with error
pub struct Quadratic {
    /// Coefficient of the quadratic term
    pub a: BigInt,
    /// Coefficient of the linear term
    pub b: BigInt,
    /// The constant term
    pub c: BigInt,
}

impl Quadratic {
    /// Create a new quadratic equation instance using asset amounts defined in the `docs/astrozap.pdf`
    pub fn from_asset_amounts(
        offer_user: &BigInt,
        offer_pool: &BigInt,
        ask_user: &BigInt,
        ask_pool: &BigInt,
    ) -> Self {
        let a = ask_pool + ask_user;

        // the 1st term of b
        let b1 = offer_pool * &a * 2;
        // the 2nd term of b
        let b2 = ask_pool * (offer_pool + offer_user) * COMMISSION_RATE_BPS / 10000;
        // combine the two terms
        let b = b1 - b2;

        let c = offer_pool * (offer_pool * ask_user - offer_user * ask_pool);

        Self { a, b, c }
    }

    /// Compute value of the function by the given x
    ///
    /// f(x) = a * x * x + b * x + c
    pub fn compute_value(&self, x: &BigInt) -> BigInt {
        &self.a * x * x + &self.b * x + &self.c
    }

    /// Compute value of the function by the given x
    ///
    /// f'(x) = 2 * a * x + b
    pub fn compute_deriv_value(&self, x: &BigInt) -> BigInt {
        2 * &self.a * x + &self.b
    }

    /// Solve the quadratic equation `a * x^2 + b * x + c = 0` using Newton's method
    ///
    /// If `MAX_ITERATION` is reached without convergence, we simply return the latest x value. The
    /// x value at this time does not represent the optimal swap amount, but it is fine because we
    /// will check slippage tolerance at the very end of the function call, so liquidity provisions
    /// with too big slippage will be reverted.
    /// 
    /// Also, in practice, almost all such equations converge in 4 - 5 iterations.
    pub fn solve(&self) -> BigInt {
        let mut x_prev: BigInt = 0.into();
        let mut x = x_prev.clone();

        // x_{n+1} = x_n - f(x_n) / f'(x_n)
        for _ in 0..MAX_ITERATIONS {
            let val = self.compute_value(&x_prev);
            let deriv_val = self.compute_deriv_value(&x_prev);
            x = &x_prev - val / deriv_val;
            if x == x_prev {
                break;
            } else {
                x_prev = x.clone()
            }
        }

        x
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::helpers::bigint_to_uint128;
    use cosmwasm_std::Uint128;

    fn mock_equation() -> Quadratic {
        Quadratic::from_asset_amounts(
            &100000000000u128.into(),
            &118070429547232u128.into(),
            &0.into(),
            &1451993415113u128.into(),
        )
    }

    #[test]
    fn should_create_from_asset_amounts() {
        let qe = mock_equation();
        assert_eq!(qe.a, BigInt::from(1451993415113u128));
        assert_eq!(qe.b, BigInt::from(342360224387597541370186081u128));
        assert_eq!(qe.c,BigInt::from(-17143748622214425401611721600000000000i128));
    }

    #[test]
    fn should_solve_equation() {
        let qe = mock_equation();
        let offer_amount_bi = qe.solve();
        let offer_amount = bigint_to_uint128(&offer_amount_bi).unwrap();
        assert_eq!(offer_amount, Uint128::new(50064546170u128));
    }
}

File: scripts/math.ts
import BN from "bn.js";

type BigNumberish = number | string | BN;

/**
 * TypeScript implementation of the `Equation` struct defined in `astrozap/src/math.rs`
 *
 * Represents a quadratic equation of form `a * x^2 + b * x - c = 0`
 */
export class Equation {
  MAX_ITERATIONS = 32;
  a: BN;
  b: BN;
  c: BN;

  /**
   * Create a new `Equation` instance using asset amounts
   *
   * See `docs/astrozap.pdf` for explaination
   */
  constructor(
    offer_user: BigNumberish,
    offer_pool: BigNumberish,
    ask_user: BigNumberish,
    ask_pool: BigNumberish
  ) {
    offer_user = new BN(offer_user);
    offer_pool = new BN(offer_pool);
    ask_user = new BN(ask_user);
    ask_pool = new BN(ask_pool);

    this.a = ask_pool.add(ask_user);

    let b1 = offer_pool.mul(this.a).mul(new BN(2));
    let b2 = ask_pool.mul(offer_pool.add(offer_user)).mul(new BN(30)).div(new BN(10000));
    this.b = b1.sub(b2);

    this.c = offer_pool.mul(offer_pool.mul(ask_user).sub(offer_user.mul(ask_pool)));
  }

  /**
   * Print the paramters a, b, c
   */
  parameters() {
    console.log("a =", this.a.toString());
    console.log("b =", this.b.toString());
    console.log("c =", this.c.toString());
  }

  /**
   * Compute the value of `f(x) = a * x^2 + b * x - c` at the given x
   */
  computeValue(x: BigNumberish) {
    x = new BN(x);
    return this.a.mul(x).mul(x).add(this.b.mul(x)).add(this.c);
  }

  /**
   * Compute the value of `f'(x) = 2 * a * x + b` at the given x
   */
  computeDerivValue(x: BigNumberish) {
    x = new BN(x);
    return new BN(2).mul(this.a).mul(x).add(this.b);
  }

  /**
   * Solve the equation `f(x) = 0` using Newton's method with the given initial value
   */
  solve() {
    let xPrev = new BN(0);
    let x = xPrev;

    // x_{n+1} = x_n - f(x_n) / f'(x_n)
    for (let i = 0; i < this.MAX_ITERATIONS; i++) {
      let val = this.computeValue(x);
      let derivVal = this.computeDerivValue(x);
      x = xPrev.sub(val.div(derivVal));

      console.log(
        `iteration ${i + 1}, x = ${x.toString()}, val = ${val.toString()}, derivVal = ${derivVal.toString()}`
      );

      if (x.toNumber() == xPrev.toNumber()) {
        break;
      } else {
        xPrev = x;
      }
    }

    return x.toString();
  }
}
